### PR TITLE
Migrating existing HuggingFace support to a more robust one

### PR DESCRIPTION
**Pull request by Anindyadeep**
Current status: WIP 

The PR motivation started from this issue: #1018 

### What this PR do?

We have an existing HuggingFace model integration on DsPY. However developers can not use this class with much flexibility. So the goal of the PR is to provide a balanced abstraction such that developers can

1. add their own model object
2. or add just different configs and arguments while instantiating, example: quantization_config, lora_config directly in the dspy.HF class
3. make it also directly compatible with the finetuning procedures (might not be applicable now)

The reason is, HuggingFace is very popular for sometimes writing out entire lifecycle. So writing entre pipelines with huggingface and dspy might need to have almost all the supported configs and HF's own addons (like bitsandbytes, peft etc), so that they can focus more on dspy instead of finding wayaround on how to add this or that with dspy huggingface.

### CODE REVIEW

**Comment by Anindyadeep** [comment]
Hi would love to get your thoughts on this, so that according to the feedbacks I can move forward. Thanks 

cc: @arnavsinghvi11

**Comment by Anindyadeep** [comment]
> Current status: WIP
> 
> The PR motivation started from this issue: #1018
> 
> ### What this PR do?
> We have an existing HuggingFace model integration on DsPY. However developers can not use this class with much flexibility. So the goal of the PR is to provide a balanced abstraction such that developers can
> 
> 1. add their own model object
> 2. or add just different configs and arguments while instantiating, example: quantization_config, lora_config directly in the dspy.HF class
> 3. make it also directly compatible with the finetuning procedures (might not be applicable now)
> 
> The reason is, HuggingFace is very popular for sometimes writing out entire lifecycle. So writing entre pipelines with huggingface and dspy might need to have almost all the supported configs and HF's own addons (like bitsandbytes, peft etc), so that they can focus more on dspy instead of finding wayaround on how to add this or that with dspy huggingface.
> 
> ### Todos
> * [x]  Initial working draft implementation
> * [x]  Testing the implementation passing HF models object
> * [x]  Testing the implementation passing HF models object as string
> * [x]  Testing the implementation passing HF models with quantization support
> * [ ]  Testing the implementation passing HF models with multi-gpu support (currently not have multi gpu settings)
> * [ ]  Refine the code (check once more, typos, check on some todos I wrote inside the code etc)
> * [ ]  Write good tutorial like documentation for this
> * [ ]  Fininshing (link etc)
> 
> cc: @okhat, let's discuss more about this approach and feel free to provide feedbacks :)

Hi just wanted to check on whether we can move forward with this PR or not, else we can close it. Thanks :)

File: dsp/modules/hf_new.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        bnb_config: Optional["transformers.BitsAndBytesConfig"]=None, 
+        **kwargs,
+    ):
+        try:
+            import torch
+            import transformers
+        except ImportError as exc:
+            raise ModuleNotFoundError(
+                "You need to install the following libraries: ",
+                "transformers >= 4.30.0, torch==2.2.2, accelerator"
```
Comment (arnavsinghvi11) [inline dsp/modules/hf_new.py:44]: would eventually be added to pyproject.toml and/or poetry?

**Comment by arnavsinghvi11** [comment]
Hi @Anindyadeep , thanks for this PR refactor of HF! This looks great!

I anticipate this PR will eventually replace [`hf.py`](https://github.com/stanfordnlp/dspy/blob/main/dsp/modules/hf.py) so it would be great to either build on the existing file if possible. It would also be easier to understand what’s being retained and any additional differences. Following the leftover todos and testing, let me know if you have any comments/additions to review.

Looking forward to the next stage of this PR!

**Comment by Anindyadeep** [comment]
> Hi @Anindyadeep , thanks for this PR refactor of HF! This looks great!
> 
> I anticipate this PR will eventually replace [`hf.py`](https://github.com/stanfordnlp/dspy/blob/main/dsp/modules/hf.py) so it would be great to either build on the existing file if possible. It would also be easier to understand what’s being retained and any additional differences. Following the leftover todos and testing, let me know if you have any comments/additions to review.
> 
> Looking forward to the next stage of this PR!

Hey thanks for the reply, on a second though I feel this PR is a very big addition so in that case, I could re-start from the PR implementation, where I can break into smaller parts (which could be easier for you all as well for review) and doing changes from `hf.py`. Does that sounds good to you?

**Comment by Anindyadeep** [comment]
> > Hi @Anindyadeep , thanks for this PR refactor of HF! This looks great!
> > I anticipate this PR will eventually replace [`hf.py`](https://github.com/stanfordnlp/dspy/blob/main/dsp/modules/hf.py) so it would be great to either build on the existing file if possible. It would also be easier to understand what’s being retained and any additional differences. Following the leftover todos and testing, let me know if you have any comments/additions to review.
> > Looking forward to the next stage of this PR!
> 
> Hey thanks for the reply, on a second though I feel this PR is a very big addition so in that case, I could re-start from the PR implementation, where I can break into smaller parts (which could be easier for you all as well for review) and doing changes from `hf.py`. Does that sounds good to you?

Let me know your thoughts @arnavsinghvi11 whenever you find some bandwidth, thanks :)

**Comment by arnavsinghvi11** [comment]
>  I could re-start from the PR implementation, where I can break into smaller parts (which could be easier for you all as well for review) and doing changes from hf.py. Does that sounds good to you?

Yes that would be great!

**Comment by okhat** [comment]
Excellent effort here! Tagging @dilarasoylu and @isaacbmiller

**Comment by okhat** [comment]
I wonder how much of this can we delegate to an opinionated tool like axotol or torchtune or something btw?

**Comment by Anindyadeep** [comment]
> I wonder how much of this can we delegate to an opinionated tool like axotol or torchtune or something btw?

Well, it depends, I mean sure, we can go for axotol or torchtune, but then it would be less coupled with native hf scripts imo. However let me know what do you think on this and we can go accordingly.

**Comment by okhat** [comment]
My only goal is to support most useful models & quantization settings & loss functions (DPO etc) with as little code as possible. Do you think HF is still the best path for that?

**Comment by okhat** [comment]
If HF is still best, let's stick to it! Thanks for the amazing work here @Anindyadeep !

**Comment by Anindyadeep** [comment]
> My only goal is to support most useful models & quantization settings & loss functions (DPO etc) with as little code as possible. Do you think HF is still the best path for that?

Ahh yes, in that case, we can go for HF, okay then let me complete this PR and I will ping you guys. (Putting it into draft now)
